### PR TITLE
Tx/Vp Whitelist

### LINKED ISSUE #875: validity predicate whitelist
Write validity predicates that implement whitelist that limits which VPs can be used to create new accounts or to update VPs of existing accounts (we can use their hashes). For this, all the validity predicates included in the whitelist must themselves perform this check to ensure that no VP that is not included in the whitelist can be deployed.

### LINKED ISSUE #876: tx whitelist
Similar to #875, we'd also like to limit which txs can be used (by tx WASM code hash). This check can also be performed by all the VPs so that no change is required in the base ledger.

### PR DESCRIPTION
**Pull request by Fraccaman**
Based on #889. Close #876. Close #875.

### CODE REVIEW

**Comment by Fraccaman** [comment]
pls update wasm

**Comment by Fraccaman** [comment]
pls update wasm

File: .changelog/unreleased/features/875-tx-whitelist.md [hunk 1]
```diff
@@ -0,0 +1 @@
+- Add tx whitelist ([#875](https://github.com/anoma/anoma/issues/875))
```
Comment (tzemanovic) [inline .changelog/unreleased/features/875-tx-whitelist.md:1]: this one should be #876

Comment (tzemanovic) [inline .changelog/unreleased/features/875-tx-whitelist.md:1]: this one should be #876

File: apps/src/lib/config/genesis.rs [hunk 2]
```diff
@@ -183,6 +183,10 @@ pub mod genesis_config {
         // Maximum duration per block (in seconds).
         // TODO: this is i64 because datetime wants it
         pub max_expected_time_per_block: i64,
+        // Hashes of whitelisted vps array
```
Comment (tzemanovic) [inline apps/src/lib/config/genesis.rs:186]:
```suggestion
        // Hashes of whitelisted vps array. `None` value or an empty array disables whitelisting.
```

File: vp_prelude/src/lib.rs [hunk 3]
```diff
@@ -4,7 +4,32 @@
...
+    // if whitelist is empty, allow any transaction
+    whitelist.contains(&tx_hash.to_string()) || whitelist.is_empty()
+}
+
+pub fn is_vp_whitelisted(vp_bytes: &[u8]) -> bool {
+    let vp_hash = sha256(vp_bytes);
+    let key = parameters::vp_whitelist_storage_key();
+    let whitelist: Vec<String> = read_pre(&key.to_string()).unwrap_or_default();
+    // if whitelist is empty, allow any transaction
+    whitelist.contains(&vp_hash.to_string()) || whitelist.is_empty()
```
Comment (tzemanovic) [inline vp_prelude/src/lib.rs:31]: same here
```suggestion
     whitelist.is_empty() || whitelist.contains(&vp_hash.to_string())
```

File: shared/src/types/hash.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                    value.len(),
+                    HASH_LENGTH
+                ),
+            });
+        }
+        let mut hash = [0; 32];
+        hash.copy_from_slice(value);
+        Ok(Hash(hash))
+    }
+}
```
Comment (tzemanovic) [inline shared/src/types/hash.rs:66]: you can use the try_from on the fixed-size array:

```rust
#[allow(missing_docs)]
#[derive(Error, Debug)]
pub enum TryFromSliceError {
    #[error("Failed trying to convert slice to a hash: {0}")]
    ConversionFailed(std::array::TryFromSliceError),
}

impl TryFrom<&[u8]> for Hash {
    type Error = TryFromSliceError;

    fn try_from(value: &[u8]) -> Result<Self, TryFromSliceError> {
        let hash: [u8; 32] = TryFrom::try_from(value)
            .map_err(|e| TryFromSliceError::ConversionFailed(e))?;
        Ok(Hash(hash))
    }
}
```

File: tests/src/e2e/ledger_tests.rs [hunk 5]
```diff
@@ -659,6 +661,8 @@ fn pos_init_validator() -> Result<()> {
                 min_num_of_blocks: 2,
                 min_duration: 1,
                 max_expected_time_per_block: 1,
+                vp_whitelist: Some(vec![]),
+                tx_whitelist: Some(vec![]),
```
Comment (tzemanovic) [inline tests/src/e2e/ledger_tests.rs:665]: (same here)
```suggestion
                ..genesis.parameters
```

File: .changelog/unreleased/features/876-tx-whitelist.md [hunk 6]
```diff
@@ -0,0 +1 @@
+- Ledger: Added transaction whitelist configurable for a network to limit which transactions are permitted ([#876](https://github.com/anoma/anoma/issues/875))
```
Comment (tzemanovic) [inline .changelog/unreleased/features/876-tx-whitelist.md:1]:
```suggestion
- Ledger: Added transaction whitelist configurable for a network to limit which transactions are permitted ([#876](https://github.com/anoma/anoma/issues/876))
```

File: .changelog/unreleased/features/876-vp-whitelist.md [hunk 7]
```diff
@@ -0,0 +1 @@
+- Ledger: Added validity predicate whitelist configurable for a network to limit which validity predicates are permitted ([#876](https://github.com/anoma/anoma/issues/875))
```
Comment (tzemanovic) [inline .changelog/unreleased/features/876-vp-whitelist.md:1]:
```suggestion
- Ledger: Added validity predicate whitelist configurable for a network to limit which validity predicates are permitted ([#875](https://github.com/anoma/anoma/issues/875))
```

**Pull request merged**